plugins/modules/purefa_pg.py

```diff
@@ -491,203 +607,328 @@ def update_pgroup(module, array):
+                                group_names=[module.params["name"]],
+                                context_names=[module.params["context"]],
+                                member_names=module.params["host"],
                             )
-                        except Exception:
+                        else:
+                            res = array.delete_protection_groups_hosts(
+                                group_names=[module.params["name"]],
+                                member_names=module.params["host"],
+                            )
+                        if res.status_code != 200:
                             module.fail_json(
-                                msg="Removing hosts in pgroup {0} failed.".format(
-                                    module.params["name"]
+                                msg="Removing hosts from pgroup {0} failed. Error: {1}".format(
+                                    module.params["name"], res.errors[0].message
                                 )
                             )
 
     if (
         module.params["hostgroup"]
-        and get_pgroup(module, array)["hosts"] is None
-        and get_pgroup(module, array)["volumes"] is None
+        and get_pgroup(module, array).host_count == 0
+        and get_pgroup(module, array).volume_count == 0
     ):
-        if get_pgroup(module, array)["hgroups"] is None:
+        if get_pgroup(module, array).host_group_count == 0:
             if not module.check_mode:
                 changed = True
-                try:
-                    array.set_pgroup(
-                        module.params["name"], hgrouplist=module.params["hostgroup"]
+                if LooseVersion(CONTEXT_API_VERSION) <= LooseVersion(api_version):
+                    res = array.post_protection_groups_host_groups(
+                        group_names=[module.params["name"]],
+                        context_names=[module.params["context"]],
+                        member_names=module.params["host"],
+                    )
+                else:
+                    res = array.post_protection_groups_host_groups(
+                        group_names=[module.params["name"]],
+                        member_names=module.params["host"],
                     )
-                except Exception:
+                if res.status_code != 200:
                     module.fail_json(
-                        msg="Adding hostgroups to pgroup {0} failed.".format(
-                            module.params["name"]
+                        msg="Adding hostgroups in pgroup {0} failed. Error: {1}".format(
+                            module.params["name"], res.errors[0].message
                         )
                     )
         else:
-            cased_hostg = list(module.params["hostgroup"])
-            cased_pghostg = list(get_pgroup(module, array)["hgroups"])
+            pghostgs = []
+            hostgs = list(
+                array.get_protection_groups_host_groups(
```

Comment:
context case handling is missing, pls add. Its there in multiple occurrences.
When running in fleet array with context, this needs param to be passed - context_names=[module.params["context"]]